src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -266,5 +275,33 @@ protected virtual string GenerateUrl(string protocol, string host, VirtualPathDa
                 builder.Clear();
             }
         }
+
+        private bool TryFastGenerateUrl(
+            string protocol, 
+            string host, 
+            VirtualPathData pathData, 
+            string fragment,
+            out string url)
+        {
+            var pathBase = HttpContext.Request.PathBase;
+            url = null;
+
+            if (string.IsNullOrEmpty(protocol) 
+                && string.IsNullOrEmpty(host) 
+                && string.IsNullOrEmpty(fragment)
+                && !pathBase.HasValue)
+            {
+                if (pathData.VirtualPath.Length == 0)
+                {
+                    url = "/";
+                }
+                else if (pathData.VirtualPath.StartsWith("/", StringComparison.Ordinal))
+                {
+                    url = pathData.VirtualPath;
+                }
```

Comment:
Easier to read if `return true` here and default path does `return false` without `url != null`